src/menu.rs

```diff
@@ -77,20 +75,16 @@ pub fn context_menu<'a>(
         .into()
     };
 
-    let mut selected_dir = 0;
-    let mut selected = 0;
-    tab.items_opt().map(|items| {
-        for item in items.iter() {
-            if item.selected {
-                selected += 1;
-                if item.metadata.is_dir() {
-                    selected_dir += 1;
-                }
-            }
+    let (selected, selected_dir) = tab.items_opt().into_iter().fold((0, 0),
+        |selections, items| {
+            let selected_iter = items.into_iter().filter(|i| i.selected);
+            let selected_count = selected_iter.clone().count();
+            let selected_dirs_count = selected_iter.filter(|i| i.metadata.is_dir()).count();
+            (selections.0 + selected_count, selections.1 + selected_dirs_count)
```

Comment:
This replaces one for loop with two

Reply:
Actually three loops, `selected_iter`, `selected_count` and `selected dirs_count`

Reply:
It'd been proven many times Rust's functional style gives compiler much wider field for optimisations than traditional imperative loops, especially if mutability is removed.

I can't run this through [godbolt](https://godbolt.org/) as they don't have `criterion` available, but my benchmark for 1000000 items, shows **2.7281 ms** for `for` loop and **471.58 µs** for `filter()/count()` iterators on my notebook PC. For 10000 difference is smaller, but still `filter()/count()` wins: 6.3825 µs / 4.7141 µs.

Style wise, I'd personally replace `tuple` with own type supporting addition to right fields and logic queries, etc.

Reply:
This now O(n^2) instead of O(n).

Reply:
Wait, are you folding over an option